### PR TITLE
239 refactor relay algorithms

### LINKED ISSUE #239: Refactor relay algorithms
- old code was too messy and doesn't match the new code structure and logic.
- modify so that assign relay and compute relay position both happen in same function

### PR DESCRIPTION
**Pull request by Raa37**

### CODE REVIEW

File: DotBot.py [hunk 1]
```diff
@@ -48,7 +48,13 @@ def __init__(self, dotBotId, x, y, floorplan):
         self.isRelay              = False
         # if DotBot has bumped
         self.hasJustBumped        = False  # needed as variable as retransmits are called by simEngine
+        # heartbeat is estimated pdr sent to orchestrator for every heartbeat period
```
Comment (twatteyne) [inline DotBot.py]: please rephrase

File: DotBot.py [hunk 2]
```diff
@@ -48,7 +48,13 @@ def __init__(self, dotBotId, x, y, floorplan):
...
+        self.heartbeat            = 1
```
Comment (twatteyne) [inline DotBot.py]: what does `1` mean? at least add comment

File: DotBot.py [hunk 3]
```diff
@@ -48,7 +48,13 @@ def __init__(self, dotBotId, x, y, floorplan):
...
+        self.heartbeatPeriod      = 10
```
Comment (twatteyne) [inline DotBot.py]: add comment that it's in seconds

File: DotBot.py [hunk 4]
```diff
@@ -48,7 +48,13 @@ def __init__(self, dotBotId, x, y, floorplan):
...
+        self.numPacketReceived    = 0
 
+        # kickoff heartbeat
+        self.simEngine.schedule(self.simEngine.currentTime() + self.heartbeatPeriod, self._heartbeatCb)
     # ======================== public ==========================================
```
Comment (twatteyne) [inline DotBot.py:60]: add empty newline

File: DotBot.py [hunk 5]
```diff
@@ -185,6 +194,22 @@ def _movementTimeoutCb(self):
         # stop movement and send notification
         self._stopAndTransmit()
 
+    def _heartbeatCb(self):
+        '''
+        send heartbeat to orchestrator to update on PDR status
+        '''
+
+        self.simEngine.schedule(self.simEngine.currentTime() + self.heartbeatPeriod, self._heartbeatCb)
```
Comment (twatteyne) [inline DotBot.py]: add comment "# schedule next heartbeat"

File: DotBot.py [hunk 6]
```diff
@@ -185,6 +194,22 @@ def _movementTimeoutCb(self):
...
+
+        # compute estimated PDR
+        self.heartbeat         = self.numPacketReceived / self.heartbeatPeriod
```
Comment (twatteyne) [inline DotBot.py]: I don't understand what this variable is

File: Orchestrator.py [hunk 7]
```diff
@@ -68,6 +70,12 @@ def __init__(self, numRobots, initX, initY):
                     'targetCell':         None,
                     # [(cx1, cy1) -> (cx2, cy2) -> .... -> (cxTarget, cyTarget)]
                     'currentPath':        [],
+                    # heartbeat of last estimated PDR from DotBot
+                    'heartbeat':          1,
```
Comment (twatteyne) [inline Orchestrator.py]: if "heartbeat" is some sort of PDR the DotBot measured, ABSOLUTELY change its name

File: Orchestrator.py [hunk 8]
```diff
@@ -184,10 +204,14 @@ def receive(self,frame):
             # DotBot bumped into first cell on path at corner
             self.cellsObstacle += [dotBot['currentPath'][0]]
 
+            # DotBot bumped into its target cell at corner
+            if ((dotBot['x'], dotBot['y']) in self._computeCellCorners(*dotBot['targetCell'])):
+                self._isCornerFrontier(dotBot['targetCell'])
```
Comment (twatteyne) [inline Orchestrator.py:209]: this code doesn't do anything?

File: Orchestrator.py [hunk 9]
```diff
@@ -245,7 +283,7 @@ def receive(self,frame):
                 # no target, no path
                 path = None
 
-            elif targetCell != dotBot['targetCell'] or not dotBot['currentPath'] or frame['hasJustBumped']:
+            elif (targetCell != dotBot['targetCell'] or not dotBot['currentPath'] or frame['hasJustBumped']):
```
Comment (twatteyne) [inline Orchestrator.py]: add parenthesis

File: Orchestrator.py [hunk 10]
```diff
@@ -747,4 +783,151 @@ def _computeHeadingSpeedMovementTimeout(self, dotBotId, path):
...
-        return (heading, speed, movementTimeout)
\ No newline at end of file
+        return (heading, speed, movementTimeout)
+
+    # === Relays
+
+    def _assignRelaysAndRelayPositionsCb(self):
+
+        log.debug('heartbeats {}'.format([db['heartbeat'] for (_, db) in self.dotBotsView.items()]))
+        self.simEngine.schedule(self.simEngine.currentTime() + 10, self._assignRelaysAndRelayPositionsCb)
```
Comment (twatteyne) [inline Orchestrator.py:796]: 10 is a magic value!! what does this code do?!?

File: Orchestrator.py [hunk 11]
```diff
@@ -747,4 +783,151 @@ def _computeHeadingSpeedMovementTimeout(self, dotBotId, path):
...
+
+        if self.relayAlgorithm   == "Recovery":
+            self._recoveryRelayPlacement()
```
Comment (twatteyne) [inline Orchestrator.py]: turn name around so we have the same prefix: `_relayPlacementRecovery()` and `_relayPlacementSelfHealing()`

File: Orchestrator.py [hunk 12]
```diff
@@ -747,4 +783,151 @@ def _computeHeadingSpeedMovementTimeout(self, dotBotId, path):
...
+            # don't place relays if no algorithm is specified
+            pass
+
+    def _recoveryRelayPlacement(self):
+
+        # first check if we need relays
+        for (dotBotId, dotBot) in self.dotBotsView.items():
+
+            # if all DotBot PDRs are above min threshold don't assign a relay
+            if dotBot['heartbeat'] > 0.7:
```
Comment (twatteyne) [inline Orchestrator.py]: magic value! define above

File: Orchestrator.py [hunk 13]
```diff
@@ -747,4 +783,151 @@ def _computeHeadingSpeedMovementTimeout(self, dotBotId, path):
...
+        else:
+            # don't place relays if no algorithm is specified
+            pass
+
+    def _recoveryRelayPlacement(self):
+
+        # first check if we need relays
+        for (dotBotId, dotBot) in self.dotBotsView.items():
+
+            # if all DotBot PDRs are above min threshold don't assign a relay
```
Comment (twatteyne) [inline Orchestrator.py]: comment is wrong, what you code does it is: "DotBot with high PDR isn't a relay"

File: Orchestrator.py [hunk 14]
```diff
@@ -747,4 +783,151 @@ def _computeHeadingSpeedMovementTimeout(self, dotBotId, path):
...
+            if dotBot['heartbeat'] > 0.7:
+                continue
+
+            # skip dotBots that are already relays
+            if dotBot['isRelay']:
+                continue
+
+            # assign dotBot as relay
```
Comment (twatteyne) [inline Orchestrator.py]: DotBot

File: Orchestrator.py [hunk 15]
```diff
@@ -747,4 +783,151 @@ def _computeHeadingSpeedMovementTimeout(self, dotBotId, path):
...
+            dotBot['isRelay']    = True
+
+            # get stored PDR history (oldest -> latest)
+            pdrHistory         = dotBot['pdrHistory']
```
Comment (twatteyne) [inline Orchestrator.py]: indent

**Review by twatteyne — CHANGES_REQUESTED** [review]

File: Orchestrator.py (hunk 7 above)
Comment (Raa37) [reply Orchestrator.py]: we agreed on the naming previously when we first discussed this in meetings. will rename to estimatedPDR

**Review by twatteyne — APPROVED** [review]

**Pull request merged**